build: update pre-push hook to run package tests on source changes

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   updates the `pre-push` git hook to run tests when the source files are modified.
-   If the `C` source or `manifest.json` is modified, the native-addons are compiled before running the tests.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/32

## Questions

> Any questions for reviewers of this pull request?

Do we need to add support for `julia`, `R` or `python`?.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
build: git pre-push hook runs tests on source change

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: missing_dependencies
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on tools/git/hooks/pre-push:
```diff
@@ -448,10 +448,114 @@ main() {
+		# Find test files for a given package directory
+		find_test_files() {
+			local pkg_dir="$1"
+			local pattern="$2"
+
+			if [[ -d "$pkg_dir/test" ]]; then
```
```suggestion
			local pattern="$2"
			if [[ -d "$pkg_dir/test" ]]; then
```

headlessNode committed:
chore: remove empty line

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented:
> Do we need to add support for julia, R or python?.

You asking specifically about when fixtures runners change, correct?

headlessNode commented:
@kgryte Uh... I was asking about the source but now that I think about it, I think I only saw `Fortran` source in some of `BLAS` packages.

But yeah, also when fixtures runners change.

kgryte commented:
> I was asking about the source but now that I think about it, I think I only saw Fortran source in some of BLAS packages.

Only Fortran and C files from the `src` folder.

> Test fixture runners.

Actually, only when the generated fixtures change, not the runners. E.g., `fixtures/data.json`. Meaning, if a JSON file changes in a fixtures file, we should rerun the tests, but not the runners themselves.

headlessNode commented:
@kgryte For testing, could you let me know a package which has all of those? i.e. JS, C, Fortran, fixtures

nvm, I `grep`ed it.

headlessNode committed:
refactor: add support for fortran and fixtures/json changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: missing_dependencies
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---
